### PR TITLE
Decoder checks

### PR DESCRIPTION
**Pull request by dcoutts**
Integer overflow checks and complete the UTF8 validity checking.

Still todo: full boundary/edge test cases that expose this class of problem and will ensure no regressions.

### CODE REVIEW

**Comment by arybczak** [comment]
Looks good 👍 I pushed two additional commits that fix compilation on x86_32 and remove a few unused functions and redundant calls. Please take a look.

**Comment by dcoutts** [comment]
Note: @arybczak is currently working on completing the proper tests for this feature, which is why we've not merged yet. Preliminary work on tests that @arybczak is working from is here https://github.com/well-typed/cborg/commit/70580cfe3c215205d07c15c4c6b5f538ac701dd7. The tests will comprehensibly cover in and out of rage values, focusing on the boundary cases.

File: cborg/src/Codec/CBOR/Magic.hs [hunk 1]
```diff
@@ -404,28 +394,13 @@ word64ToWord (W64# w64#) =
...
-
-{-# INLINE int8ToInt #-}
-{-# INLINE int16ToInt #-}
-{-# INLINE int32ToInt #-}
-{-# INLINE int64ToInt #-}
-
-word8ToInt  (W8#  w) = I# (word2Int# w)
-word16ToInt (W16# w) = I# (word2Int# w)
+word8ToInt  (W8#  w#) = I# (word2Int# w#)
+word16ToInt (W16# w#) = I# (word2Int# w#)
```
Comment (dcoutts) [inline cborg/src/Codec/CBOR/Magic.hs]: I think we missed the inline pragmas here. Not that it matters much, these are so small they'll always be inlined.

File: cborg/src/Codec/CBOR/Read.hs [hunk 2]
```diff
@@ -1395,7 +1395,7 @@ isInt64Canonical sz i#
 
 {-# INLINE tryConsumeWord #-}
 tryConsumeWord :: Word8 -> ByteString -> DecodedToken Word
-tryConsumeWord hdr !bs = case word8ToWord hdr of
+tryConsumeWord hdr !bs = case hdr of
```
Comment (dcoutts) [inline cborg/src/Codec/CBOR/Read.hs]: No no no. This is actually essential. Drop this whole patch. Sorry.

These conversion to `Word` are essential because GHC handles `case` on `Int` and `Word` using jump tables, but all the other `Num` types like `Int8`, `Word8` etc it uses massive chains of `(==)` tests. So the conversion to `Word` is essential for good performance here, O(1) vs O(n).

Comment (dcoutts) [reply cborg/src/Codec/CBOR/Read.hs]: See also https://github.com/well-typed/cborg/blob/master/cborg/src/Codec/CBOR/Read.hs#L1429

**Comment by arybczak** [comment]
@dcoutts force pushed to get rid of the patch that removed "redundant" word8ToWord calls and added tests.

**Comment by arybczak** [comment]
@dcoutts Hmm, CI failed, but it's only for GHC 7.10.3 and with a message `The job exceeded the maximum time limit for jobs, and has been terminated.`.

**Comment by arybczak** [comment]
@dcoutts I added utf8 test, but without use of GeneralizedUTF8 as I'm not sure why do we want to use it in this case. I think the way I did it is enough (let me know if it's not). I also didn't test all 2-splits yet as that requires moving some functions to separate util module, so let's discuss the test first.

**Comment by dcoutts** [comment]
The idea with GeneralizedUTF8 would be as a way to generate "interesting" valid and invalid cases. But perhaps we don't really need interesting cases, since we're not testing a UTF8 decoder we're testing that we handle the errors from an existing decoder ok in all cases. So perhaps completely random generation (as you're doing) ok here. We do need to check that we get enough of both cases.

We should do the "all 2-splits" thing though. That's the critical thing for the UTF8 testing.

File: serialise/tests/Tests/Boundary.hs [hunk 3]
```diff
@@ -76,11 +80,10 @@ boundaryTest
...
   -> B a
-  -> Property
-boundaryTest dec a =
-  counterexample (show a') $ if outsideRange
-                             then isLeft  a'
-                             else isRight a'
+  -> Bool
+boundaryTest dec a = if outsideRange
+                     then isLeft  a'
+                     else isRight a'
```
[the remark this answers is not in the document]
Comment (arybczak) [reply serialise/tests/Tests/Boundary.hs]: It's redundant, the counterexample is shown without it.

File: serialise/tests/Tests/Boundary.hs [hunk 4]
```diff
@@ -200,6 +201,27 @@ stringBytesBoundaryTest =
...
+instance Arbitrary MaybeText where
+  arbitrary = MaybeText . BS.pack <$> arbitrary
+
+-- | Test that decoding of both valid and invalid CBOR strings produces output
+-- without exceptions hidden within.
+utf8DecodingTest :: MaybeText -> Bool
+utf8DecodingTest (MaybeText bs) =
+  case (T.decodeUtf8' bs, deserialiseFromBytes decodeString s) of
+    v@(Right _, Right _) -> deepseq v True
+    v@(Left _,  Left _)  -> deepseq v True
```
[the remark this answers is not in the document]
Comment (arybczak) [reply serialise/tests/Tests/Boundary.hs]: Good idea. I also added it to boundary tests.

File: serialise/tests/Tests/UTF8.hs [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+  deriving Show
+instance Arbitrary MaybeText where
+  arbitrary = MaybeText . BS.pack <$> arbitrary
+
+-- | Test that decoding of both valid and invalid CBOR strings produces output
+-- without exceptions hidden within.
+utf8DecodingTest :: MaybeText -> Property
+utf8DecodingTest (MaybeText bs) = case T.decodeUtf8' bs of
+  Right _ -> collect "valid utf8"   $     (and splitsOk)
+  Left  _ -> collect "invalid utf8" $ not (or splitsOk)
```
Comment (dcoutts) [inline serialise/tests/Tests/UTF8.hs]: :+1:
